### PR TITLE
termusic: 0.7.10 -> 0.7.11

### PR DESCRIPTION
**Pull request by VuiMuich**
## Description of changes

[Change log](https://github.com/tramhao/termusic/blob/v0.7.11/CHANGELOG.md#v0711)
Most notable change is that the actual player is split of into a `termusi-server` that is able to run in the background and comes as its own binary and the tui/launcher `termusic` binary.
This split made it necessary to switch the `fetcher` to pull directly the tarball from github, as on crates.io those are two distinct and independent crates.

The one thing where I am not sure is, if `darwin` would need the openssl dep explicitly as well.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/audio/termusic/default.nix [hunk 1]
```diff
@@ -1,31 +1,38 @@
...
   pname = "termusic";
-  version = "0.7.10";
+  version = "v0.7.11";
 
-  src = fetchCrate {
-    inherit pname version;
-    hash = "sha256-m0hi5u4BcRcEDEpg1BoWXc25dfhD6+OJtqSZfSdV0HM=";
+  src = fetchFromGitHub {
+    owner = "tramhao";
+    repo = pname;
```
Comment (drupol) [inline pkgs/applications/audio/termusic/default.nix:19]: `repo = pname;` is nice for DRY but creates a binding that goes too far.

See further information about this here: https://github.com/nix-community/nixpkgs-lint/issues/21

**Review by drupol — CHANGES_REQUESTED** [review]
Hi !

Thanks for your contribution.

I've left some comments, left me know if this is clear enough.

Don't forget to update the PR title (see contributing guide).

**Comment by VuiMuich** [comment]
Thanks for the review.
The requested change totally makes sense.
May I just ask what "DRY" refers to? I couldn't find the meaning in the linked issue or the PR the discussion arose from.

Sorry for the messed title, I was sure I put it right in the first place. However, now it's correct, right?

File: pkgs/applications/audio/termusic/default.nix (hunk 1 above)
Comment (drupol) [reply pkgs/applications/audio/termusic/default.nix]: DRY = Don't Repeat Yourself

File: pkgs/applications/audio/termusic/default.nix [hunk 2]
```diff
@@ -1,31 +1,38 @@
...
-  version = "0.7.10";
+  version = "v0.7.11";
 
-  src = fetchCrate {
-    inherit pname version;
-    hash = "sha256-m0hi5u4BcRcEDEpg1BoWXc25dfhD6+OJtqSZfSdV0HM=";
+  src = fetchFromGitHub {
+    owner = "tramhao";
+    repo = "termusic";
+    rev = version;
```
Comment (ius) [inline pkgs/applications/audio/termusic/default.nix:20]:
```nix
rev = "v${version}";
```

File: pkgs/applications/audio/termusic/default.nix [hunk 3]
```diff
@@ -1,31 +1,38 @@
...
+, openssl
 , pkg-config
 , alsa-lib
 , darwin
 }:
 
 rustPlatform.buildRustPackage rec {
   pname = "termusic";
-  version = "0.7.10";
+  version = "v0.7.11";
```
Comment (ius) [inline pkgs/applications/audio/termusic/default.nix:15]:
```nix
version = "0.7.11";
```

**Review by ius — CHANGES_REQUESTED** [review]

File: pkgs/applications/audio/termusic/default.nix (hunk 3 above)
Comment (drupol) [reply pkgs/applications/audio/termusic/default.nix:15]: ?

Comment (ius) [reply pkgs/applications/audio/termusic/default.nix:15]: The `v` is part of the GitHub tag and shouldn't end up in the `version` attribute itself...

Comment (drupol) [reply pkgs/applications/audio/termusic/default.nix:15]: Ok got it. I fully agree.

Comment (VuiMuich) [reply pkgs/applications/audio/termusic/default.nix]: Thanks for this hint, I actually disliked adding that 'v' there, but didn't see the possibility to expand the variable inside the string.

File: pkgs/applications/audio/termusic/default.nix [hunk 4]
```diff
@@ -1,31 +1,38 @@
...
-  version = "0.7.10";
+  version = "0.7.11";
 
-  src = fetchCrate {
-    inherit pname version;
-    hash = "sha256-m0hi5u4BcRcEDEpg1BoWXc25dfhD6+OJtqSZfSdV0HM=";
+  src = fetchFromGitHub {
+    owner = "tramhao";
+    repo = "termusic";
+    rev = "v$(version)";
```
Comment (drupol) [inline pkgs/applications/audio/termusic/default.nix:20]:
```suggestion
    rev = "v${version}";
```

**Review by drupol — CHANGES_REQUESTED** [review]

**Review by drupol — APPROVED** [review]
The changes LGTM, let's wait for the CI and merge this.

**Comment by VuiMuich** [comment]
Thanks for the review and patience with my noob mistakes :pray:

**Comment by drupol** [comment]
> Thanks for the review and patience with my noob mistakes pray

It's perfectly fine, we are here for helping newcomers :)

**Review by ius — APPROVED** [review]

File: pkgs/applications/audio/termusic/default.nix [hunk 5]
```diff
@@ -1,31 +1,38 @@
...
-    hash = "sha256-m0hi5u4BcRcEDEpg1BoWXc25dfhD6+OJtqSZfSdV0HM=";
+  src = fetchFromGitHub {
+    owner = "tramhao";
+    repo = "termusic";
+    rev = "v${version}";
+    hash = "sha256-ykOBXM/WF+zasAt+6mgY2aSFCpGaYcqk+YI7YLM3MWs=";
   };
 
-  cargoHash = "sha256-A83gLsaPm6t4nm7DJfcp9z1huDU/Sfy9gunP8pzBiCA=";
+  cargoHash = "sha256-1Itrq4O/sCJNHNU1NrfKhCMb+JJYGfq/vsrMXvWDdJU=";
```
Comment (drupol) [inline pkgs/applications/audio/termusic/default.nix:24]:
```suggestion
  cargoHash = "sha256-BrOpU0RFdlRXQIMjfHfs/XYIdBCYKFSA+5by/rGzC8Y=";
```

**Review by drupol — CHANGES_REQUESTED** [review]

**Review by drupol — CHANGES_REQUESTED** [review]
Could you please squash all your commits into one? This PR should only contains 1 single commit in the end.

Thanks!

**Comment by VuiMuich** [comment]
Sure, will do in a minute.

What I don't really get is, what caused the `cargo-hash` to change?

**Comment by drupol** [comment]
> Sure, will do in a minute.
> 
> What I don't really get is, what caused the `cargo-hash` to change?

I don't really know. Maybe someone doing rust might help here (@figsoda maybe?).

**Comment by drupol** [comment]
@GrahamcOfBorg build termusic

**Comment by drupol** [comment]
@GrahamcOfBorg build termusic

**Comment by drupol** [comment]
@GrahamcOfBorg build termusic

File: pkgs/applications/audio/termusic/default.nix [hunk 6]
```diff
@@ -12,6 +12,7 @@
 , rustPlatform
 , sqlite
 , stdenv
+, SystemConfiguration
```
Comment (drupol) [inline pkgs/applications/audio/termusic/default.nix:15]: This is not needed since we're passing `darwin`.

**Comment by VuiMuich** [comment]
I copied the dependency from the `atuin` package as I am pretty sure it builds well on Darwin.

File: pkgs/applications/audio/termusic/default.nix [hunk 7]
```diff
@@ -47,6 +48,7 @@ rustPlatform.buildRustPackage rec {
     darwin.apple_sdk.frameworks.IOKit
     darwin.apple_sdk.frameworks.MediaPlayer
     darwin.apple_sdk.frameworks.Security
+    SystemConfiguration
```
Comment (drupol) [inline pkgs/applications/audio/termusic/default.nix:51]:
```suggestion
    darwin.apple_sdk.frameworks.SystemConfiguration
```

**Review by drupol — CHANGES_REQUESTED** [review]

**Comment by drupol** [comment]
@GrahamcOfBorg build termusic

File: pkgs/applications/audio/termusic/default.nix (hunk 6 above)
Comment (VuiMuich) [reply pkgs/applications/audio/termusic/default.nix:15]: :thinking: but x86_64 "failed" neutral, right?

File: pkgs/applications/audio/termusic/default.nix (hunk 7 above)
Comment (drupol) [reply pkgs/applications/audio/termusic/default.nix]: I fixed it myself.

File: pkgs/applications/audio/termusic/default.nix (hunk 6 above)
Comment (drupol) [reply pkgs/applications/audio/termusic/default.nix]: I fixed it myself.

**Comment by drupol** [comment]
@GrahamcOfBorg build termusic

File: pkgs/applications/audio/termusic/default.nix (hunk 6 above)
Comment (drupol) [reply pkgs/applications/audio/termusic/default.nix]: Let's see how it goes now.

File: pkgs/applications/audio/termusic/default.nix (hunk 7 above)
Comment (VuiMuich) [reply pkgs/applications/audio/termusic/default.nix]: Thanks, I was unsure if this prefix was as needed, but thought the other package build well without it.. sorry for the fuzz.

Comment (drupol) [reply pkgs/applications/audio/termusic/default.nix]: No problem :)

**Comment by drupol** [comment]
@GrahamcOfBorg build termusic

**Comment by drupol** [comment]
Looks like it works now.

When everything is ok, I'll squash/merge the PR.

**Comment by VuiMuich** [comment]
:tada: thanks again @drupol for your welcoming and kind support on my first contribution!
This is the sustainable way to ensure quality in OSS and the foundation for great communities.

**Comment by drupol** [comment]
Welcome to the NixOS community :)

Keep PRs coming!

**Comment by VuiMuich** [comment]
I will, and the next will probably be linux only :grin:

**Comment by devhell** [comment]
Sorry for the late response. This is awesome work, @VuiMuich! Also, thank you to @ius and especially @drupol for helping with this. Thanks everyone, prost! :beers:

**Comment by VuiMuich** [comment]
> Sorry for the late response. This is awesome work, @VuiMuich! Also, thank you to @ius and especially @drupol for helping with this. Thanks everyone, prost! :beers: 

Much appreciated, it was a great experience for a first contribution on such a huge project!
Prost 🍻

**Pull request merged**